### ISSUE TITLE
Feature suggestion: Iterating over part of the topological order in the DirectedAcyclicGraph

### ISSUE THREAD

**Issue by kevin-dp** [jgrapht/jgrapht#899]
Currently, the `DirectedAcyclicGraph` provides an `iterator` method to iterate over the topological order. Since the algorithm is dynamic, the user may be interested only in the part that changed. Currently, it is not possible to iterate only over the part that changed.

When the user modifies the graph, we could return the start and end index of the affected region (which is already kept internally as `affectedRegion`). We could also provide users with a method to fetch the vertex at a given position in the topological order. If we do this, the indices must correspond to the actual indices of those vertices in the topological order. Currently, this is not the case since the internal `topoOrderMap` may contain holes.

_**Example:**_

After implementing this feature, users could iterate over the affected region like this:

```
Region reg = graph.addEdge(from, to); // region affected by this edge
if (reg != null) {
  for (var i = reg.getStart(); i <= reg.getFinish(); i++) {
    graph.getVertex(i);
  }
}
```

_**Changes required by this proposal:**_

1. Add public methods to fetch the start and end of the affected region.
2. Let the `addEdge`, `addVertex`, and `removeVertex` methods not only return a boolean but also the affected region. E.g. return the affected region if successful, and null otherwise.
3. Add public method to fetch the vertex at a given index in the topological order.
4. Modify the implementations of `TopoOrderMap` as to remove holes. The major change here is to let `removeVertex` shift all vertices that follow the removed vertex, one position to the left.

I already implemented changes 1, 3, and 4. These can be found in my [local fork](https://github.com/jgrapht/jgrapht/compare/master...kevin-dp:master). If this is interesting, I can implement change 2 and make a pull request.

**_Performance impact:_**

Currently, the `iterator()` method loops over the `topoOrderMap` thereby skipping over holes.
Hence, we pay the cost of these holes every time we iterate over the topological order.
Modifying the `removeVertex` operation from `TopoOrderMap` as explained in change 4, ensures we pay this cost only once, namely, when removing the vertex.

The method added in change 3 returns the vertex at a given position in O(1) since it can fetch it from the internal `topoOrderMap`.

**Comment by d-michail** [comment]
The use-case sounds reasonable, however your suggested implementation is trying to expose internal representation. In Java and in other object oriented languages, iterators is the proper way to go when you want to iterate over something. 

For point 1 you should create a method which returns an iterator which implements `Iterator<V>` and iterates over vertices in the (last) affected region. Maybe you can call the method `lastAffectedRegionIterator()` or something similar. The implementation of the iterator should be a private inner class, not visible to the user.

The methods in your second suggestion are overloading methods from the `Graph<V,E>` interface, which is the library's  main interface and cannot change.

Point 3 is no longer needed if you implement this as an iterator.

Regarding point 4, i.e. holes, we are aware of the trade-offs between iteration and removal. I would be against changing our current implementation for a single use-case. Our current implementation is stable and well-tested.

**Comment by kevin-dp** [comment]
The `lastAffectedRegionIterator()` method sounds like a good idea.
I believe we will need some extra method `lastAffectedRegionPosition()` to know where in the topological order this iterator starts iterating. Otherwise, i don't see the advantage of having a `lastAffectedRegionIterator()` if you don't know where in the topological order it is iterating over. But then, this requires removing the holes such that the position corresponds to the actual index in the topological order...

These problems probably come from the fact that i need a dynamic vector of topologically sorted vertices rather than an iterator. That's why i'm also interested in the position of the `lastAffectedRegionIterator` in the topologically sorted sequence of vertices.
Would it make sense to leave the iterator unchanged and introduce an additional topologically sorted vector that is updated incrementally?

**Comment by d-michail** [comment]
Iterators are objects which hide the actual implementation. They maintain internally the actual index, they skip over holes, etc. and when you call `getNext()` they return the next vertex in the order. 

This is exactly what happens with the `TopoIterator()`. It maintains the index of its location, it skips holes, etc. and when you ask for the next vertex using `getNext()` it returns the next vertex.
